package client;

enum Type {
  SYNC = 0;
  SET = 1;
  GET = 2;
  DEL = 3;
  INFOSTATS = 4;
  INFOCAPACITY = 5;
  INFOPARTITION = 6;
}

enum SyncType {
  CMD = 0;
  SKIP = 1;
}

enum StatusCode {
  kOk = 0;
  kNotFound = 1;
  kWait = 2;
  kError = 3;
  kFallback = 4;
}

message Node {
  required string ip = 1;
  required int32 port = 2;
}

message SyncOffset {
  required int32 filenum = 1;
  required int64 offset = 2;
  optional int32 partition = 3;
}

message CmdRequest {
  required Type type = 1;

  // Sync
  message Sync {
    required Node node = 1;
    required string table_name = 2;
    required SyncOffset sync_offset = 3;
  }
  optional Sync sync = 2; 

  message Set {
    required string table_name = 1;
    required string key = 2;
    required bytes value = 3;
    optional string uuid = 4;
  }
  optional Set set = 3;

  message Get {
    required string table_name = 1;
    required string key = 2;
    optional string uuid = 3;
  }
  optional Get get = 4;

  // Delete
  message Del {
    required string table_name = 1;
    required string key = 2;
    optional string uuid = 3;
  }
  optional Del del = 5;

  message Info {
    optional string table_name = 1; 
  }
  optional Info info = 6;

}

message CmdResponse {
  required Type type = 1;
  required StatusCode code = 2;
  optional string msg = 3;

  message Sync {
    required string table_name = 1;
    required SyncOffset sync_offset = 2;
  }
  optional Sync sync = 4;

  message Get {
    optional bytes value = 1;
  }
  optional Get get = 5;
  
  optional Node redirect = 6;
  
  // InfoStats
  message InfoStats {
    required string table_name = 1;
    required int64 total_querys = 2;
    required int32 qps = 3;
  }
  repeated InfoStats info_stats = 7;

  // InfoCapacity
  message InfoCapacity {
    required string table_name = 1;
    required int64 used = 2;
    required int64 remain = 3;
  }
  repeated InfoCapacity info_capacity = 8;

  // InfoPartition
  message InfoPartition {
    required string table_name = 1;
    repeated SyncOffset sync_offset = 2; 
  }
  repeated InfoPartition info_partition = 9;
}

message BinlogSkip {
  required string table_name = 1; 
  required int32 partition_id = 2; 
  required int64 gap = 3;
}

message SyncRequest {
  required SyncType sync_type = 1;
  required int64 epoch = 2;
  required Node from = 3;
  required SyncOffset sync_offset = 4;
  optional CmdRequest request = 5;
  optional BinlogSkip binlog_skip = 6;
}
